crates/daphne/src/hpke.rs

```diff
@@ -437,6 +438,135 @@ impl From<HpkePublicKeySerde> for HpkePublicKey {
+            into_bytes(
+                match self.version {
+                    DapVersion::Draft09 => CTX_INPUT_SHARE_DRAFT09,
+                    DapVersion::Latest => CTX_INPUT_SHARE_DRAFT13,
+                },
+                DapRole::Client,
+                self.receiver.into(),
+            )
+        }
+
+        fn aad_bytes(&self) -> Result<Vec<u8>, CodecError> {
+            let mut aad = Vec::with_capacity(
+                self.task_id.encoded_len().unwrap_or_default()
+                    + self
+                        .report_metadata
+                        .encoded_len_with_param(&self.version)
+                        .unwrap_or_default()
+                    + self.public_share.len(),
+            );
+            self.task_id.encode(&mut aad)?;
+            self.report_metadata
+                .encode_with_param(&self.version, &mut aad)?;
+            encode_u32_bytes(&mut aad, self.public_share)?;
+            Ok(aad)
+        }
+    }
+
+    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    pub struct AggregateShare<'s> {
+        // info
+        pub version: DapVersion,
+        pub sender: DapAggregatorRole,
+        // aad
+        pub task_id: &'s TaskId,
+        pub agg_param: &'s DapAggregationParam,
+        pub batch_selector: &'s BatchSelector,
+    }
+
+    // https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-4.7.4-2
+    impl InfoAndAad for AggregateShare<'_> {
+        fn info_bytes(&self) -> Vec<u8> {
+            into_bytes(
+                match self.version {
+                    DapVersion::Draft09 => CTX_AGG_SHARE_DRAFT09,
+                    DapVersion::Latest => CTX_AGG_SHARE_DRAFT13,
+                },
+                self.sender.into(),
+                DapRole::Collector,
+            )
+        }
+
+        fn aad_bytes(&self) -> Result<Vec<u8>, CodecError> {
+            let mut aad = Vec::with_capacity(
+                self.task_id.encoded_len().unwrap_or_default()
+                    + self.agg_param.encoded_len().unwrap_or_default()
+                    + self
+                        .batch_selector
+                        .encoded_len_with_param(&self.version)
+                        .unwrap_or_default(),
+            );
+            self.task_id.encode(&mut aad)?;
+            encode_u32_prefixed(self.version, &mut aad, |_version, bytes| {
+                self.agg_param.encode(bytes)
+            })?;
+            self.batch_selector
+                .encode_with_param(&self.version, &mut aad)?;
+
+            Ok(aad)
+        }
+    }
+
+    fn into_bytes(constant: &'static [u8], sender: DapRole, receiver: DapRole) -> Vec<u8> {
+        static ROLE_2_ID: [u8; 4] = {
+            let mut ids = [0; 4];
+            ids[DapRole::Collector as usize] = 0;
+            ids[DapRole::Client as usize] = 1;
+            ids[DapRole::Leader as usize] = 2;
+            ids[DapRole::Helper as usize] = 3;
+            ids
+        };
```

Comment:
it would be better if `DapRole` has a method `getID` that returns its id.

Reply:
replaced the table with this
```rs
        fn role_to_id(role: DapRole) -> u8 {
            match role {
                DapRole::Collector => 0,
                DapRole::Client => 1,
                DapRole::Leader => 2,
                DapRole::Helper => 3,
            }
        }
```

Reply:
This looks better and simpler